aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,502 @@
+				constraintCount += 1
+			}
+		}
+	}
+
+	if constraintCount > 1 {
+		return false
+	}
+	return true
+}
+
+func expandKmsGrantConstraints(configured *schema.Set) *kms.GrantConstraints {
+	if len(configured.List()) < 1 {
+		return nil
+	}
+
+	var constraint kms.GrantConstraints
+
+	for _, raw := range configured.List() {
+		data := raw.(map[string]interface{})
+		if contextEq, ok := data["encryption_context_equals"]; ok {
+			constraint.SetEncryptionContextEquals(stringMapToPointers(contextEq.(map[string]interface{})))
+		}
+		if contextSub, ok := data["encryption_context_subset"]; ok {
+			constraint.SetEncryptionContextSubset(stringMapToPointers(contextSub.(map[string]interface{})))
+		}
+	}
+
+	return &constraint
+}
+
+func sortStringMapKeys(m map[string]*string) []string {
+	keys := make([]string, 0)
+	for k := range m {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	return keys
+}
+
+// NB: For the constraint hash to be deterministic the order in which
+// print the keys and values of the encryption context maps needs to be
+// determistic, so sort them.
+func sortedConcatStringMap(m map[string]*string, sep string) string {
+	var strList []string
+	mapKeys := sortStringMapKeys(m)
+	for _, key := range mapKeys {
+		strList = append(strList, key, *m[key])
+	}
+	return strings.Join(strList, sep)
+}
+
+// The hash needs to encapsulate what type of constraint it is
+// as well as the keys and values of the constraint.
+func resourceKmsGrantConstraintsHash(v interface{}) int {
+	var buf bytes.Buffer
+	m, castOk := v.(map[string]interface{})
+	if !castOk {
+		return 0
+	}
+
+	if v, ok := m["encryption_context_equals"]; ok {
+		if len(v.(map[string]interface{})) > 0 {
+			buf.WriteString(fmt.Sprintf("encryption_context_equals-%s-", sortedConcatStringMap(stringMapToPointers(v.(map[string]interface{})), "-")))
+		}
+	}
+	if v, ok := m["encryption_context_subset"]; ok {
+		if len(v.(map[string]interface{})) > 0 {
+			buf.WriteString(fmt.Sprintf("encryption_context_subset-%s-", sortedConcatStringMap(stringMapToPointers(v.(map[string]interface{})), "-")))
+		}
+	}
+
+	return hashcode.String(buf.String())
+}
+
+func flattenKmsGrantConstraints(constraint *kms.GrantConstraints) *schema.Set {
+	constraints := schema.NewSet(resourceKmsGrantConstraintsHash, []interface{}{})
+	if constraint == nil {
+		return constraints
+	}
+
+	m := make(map[string]interface{}, 0)
+	if constraint.EncryptionContextEquals != nil {
+		if len(constraint.EncryptionContextEquals) > 0 {
+			m["encryption_context_equals"] = pointersMapToStringList(constraint.EncryptionContextEquals)
+		}
+	}
+	if constraint.EncryptionContextSubset != nil {
+		if len(constraint.EncryptionContextSubset) > 0 {
+			m["encryption_context_subset"] = pointersMapToStringList(constraint.EncryptionContextSubset)
+		}
+	}
+	constraints.Add(m)
+
+	return constraints
+}
+
+// Custom error, so we don't have to rely on
+// the content of an error message
+type KmsGrantMissingError string
```

Comment:
Instead of creating our own error class, can we instead utilize the upstream SDK's `awserr.Error` and set to a code like `kms.ErrCodeNotFoundException`?

Reply:
`kms.ErrCodeNotFoundException` is a string, and I wanted a type or code to use so I don't have to key off the error message. Might be missing something here?